shumai/tensor/tensor.ts

```diff
@@ -214,13 +232,15 @@ export function backward(base_t: Tensor, jacobian: Tensor) {
 
   // NB: can't easily embed this in the traverse functions
   // (or else references are stale)
-  const calc_grads = (all_grads_dict) => {
+  const calc_grads = (
+    all_grads_dict: Record<number, [Tensor, Tensor]>
+  ): Record<string, { grad: Tensor; tensor: Tensor }> => {
     const all_grads: Record<string, { grad: Tensor; tensor: Tensor }> = {}
     for (const key in all_grads_dict) {
       const [t, g] = all_grads_dict[key]
       // NB: not really safe in parallel, but convenient for stateful API
       t.grad = g
-      all_grads[t] = {
+      all_grads[t.ptr] = {
```

Comment:
why's this changed?  I believe `t` will be implicitly converted toString() (which uses the ptr)

Reply:
VS Code was not happy about the implicit conversion, so I thought I could make it clearer by making it explicit. It did just occur to me that I've made it a `number` instead of `string`; not sure what I was thinking.

I could restore it back to just `t`, or make it an explicit string conversion instead?

Reply:
Come to think of it, why not index it with its address like in `all_grads_dict`? Although I am not sure what the output object here gets used for.

Reply:
good question! the idea is to return a map from Tensors to gradients:

```javascript
const grad_map = loss.backward()
const grad_i_care_about = grad_map[weight].grad
```

This is to allow two different backward passes to run at the same time without both overriding `weight.grad` (an issue with the PyTorch approach)

Reply:
I think your best bet is to make it an explicit conversion to string to make VSCode happy